### PR TITLE
Handle JSON strings containing non-BMP characters

### PR DESCRIPTION
**Pull request by rhowe**
This extends the JSON string parsing to handle Unicode characters outside the Basic Multilingual Plane when they are represented using two UTF-16 surrogate pairs encoded using `\u` escapes.

Fixes https://github.com/google/jsonnet/issues/897

### CODE REVIEW

File: core/string_utils.cpp [hunk 1]
```diff
@@ -60,6 +60,50 @@ UString jsonnet_string_escape(const UString &str, bool single)
...
+        }
+        codepoint *= 16;
+        codepoint += digit;
+    }
+    return codepoint;
+}
+
+bool is_bmp_codepoint(const unsigned long codepoint)
+{
+    return codepoint != 0xd83d;
```
Comment (sbarzowski) [inline core/string_utils.cpp:93]: This is insufficient. E.g. 0xd834 is also outside bmp. IIUC 0x0000 to 0xD7FF and 0xE000 to 0xFFFF are BMP characters and everything else is non-BMP.

File: test_suite/unicode_bmp1.json [hunk 2]
```diff
@@ -0,0 +1 @@
+["\uD83D\uDCD8"]
```
Comment (sbarzowski) [inline test_suite/unicode_bmp1.json:1]: 1) It should be `unicode_bmp1.jsonnet`, not `.json`. It's parsed and treated as Jsonnet, so `.json` extension is a bit misleading here.
2) We'll need more examples like that which exercise our BMP-detection.

**Comment by sbarzowski** [comment]
Thanks! This seems pretty close to what we need. Would you mind signing the CLA, so we can use this?

File: core/string_utils.cpp (hunk 1 above)
Comment (rhowe) [reply core/string_utils.cpp:93]: Yes, I was really just raising this as a hacky demo of the kinds of changes I thought might be needed. The way I split out the \u escape parsing is pretty vile.

**Comment by rhowe** [comment]
@googlebot I signed it!

**Comment by rhowe** [comment]
OK, what was a mere demo of the bug is now I think an actual functional implementation.

However there are issues which would make me wary about it being merged in its present state:

* I do not consider myself adept at C++ - far from it. Please review the code thoroughly and don't hesitate to rewrite it.
* The hacks I did to the parser to deal with the `\u` escapes are pretty horrible and I think someone more familiar with the parser should take a look and rework them.
* The test cases are obviously weak.

**Pull request merged**